Add push GitHub Action for pushing Flux OCI artifacts

matheuscscp opened the pull request:
Introduce a GitHub Action for making it easier to push OCI artifacts with the `flux` CLI.

Tested here: https://github.com/matheuscscp/d1-fleet/actions/workflows/test-push.yaml

Closes #167

stefanprodan commented on actions/push/action.yaml:
```diff
@@ -0,0 +1,168 @@
+      echo "Pushing $INPUT_PATH to ${artifact_url}..."
+
+      push_output=$(flux push artifact oci://$artifact_url \
+        --source="$(git config --get remote.origin.url)" \
```
This would work for GHCR, we need to set it to the HTTPS URL for GitHub to link it to the repo like this https://github.com/controlplaneio-fluxcd/flux-operator/blob/cc6818a7e89ea2825ef5fa718b0da5fe94947dc5/.github/workflows/push-manifests.yml#L50

Maybe we should allow the source URL as input.

matheuscscp replied:
Yes, it works:

```json
  "labels": {
    "org.opencontainers.image.revision": "main@sha1:9cf989cad1e7ef94256c681e4b9733dd3197d063",
    "org.opencontainers.image.source": "https://github.com/matheuscscp/d1-fleet",
    "org.opencontainers.image.created": "2025-02-28T17:25:35Z"
  }
```

But I added `source` as a parameter anyway and added an example. GitHub Actions clones the repo with HTTPS, not SSH.

stefanprodan commented on actions/push/action.yaml:
```diff
@@ -0,0 +1,168 @@
+
+      push_output=$(flux push artifact oci://$artifact_url \
+        --source="$(git config --get remote.origin.url)" \
+        --revision="$(git branch --show-current)@sha1:$(git rev-parse HEAD)" \
```
Would this work with tags? Don't we need to figure out if the checkout is for a tag and do `$(git tag --points-at HEAD)@sha1:$(git rev-parse HEAD)`

matheuscscp replied:
I added this as the parameter `revision`.

stefanprodan commented on actions/push/action.yaml:
```diff
@@ -0,0 +1,168 @@
+        --output json)
+
+      digest=$(echo $push_output | jq -r .digest)
+      digest_url=$(echo $push_output | jq -r '. | .repository + "@" + .digest')
```
Do we capture errors and stop execution, I don't see pipe fail flag

matheuscscp replied:
Fixed, thanks!

stefanprodan commented on actions/push/README.md:
```diff
@@ -0,0 +1,87 @@
+      with:
+        repository: ghcr.io/${{ github.repository }}
+        path: clusters
+        source: ${{ steps.inputs.outputs.source }}
+        revision: ${{ steps.inputs.outputs.revision }}
```
This shouldn't be required, we should be able to generate these settings and avoid users messing up the revision format.

matheuscscp replied:
Fixed, thanks!

matheuscscp commented:
Everything retested after addressing all comments, this are the generated labels:

```json
  "labels": {
    "org.opencontainers.image.source": "https://github.com/matheuscscp/d1-fleet",
    "env": "production",
    "cluster": "production",
    "app": "my-app",
    "org.opencontainers.image.created": "2025-02-28T21:16:28Z",
    "org.opencontainers.image.revision": "refs/heads/main@sha1:aff6f2ea422a0c3ee21455d741fe281abdabe3d6"
  }
```

The source and revision are generated like this:

```yaml
  - name: Pre-process inputs
    id: inputs
    shell: bash
    env:
      SOURCE: ${{ github.repositoryUrl }}
      REVISION: ${{ github.ref }}@sha1:${{ github.sha }}
      ...
    run: |
      set -euo pipefail

      source=$(echo $SOURCE | sed -e 's|^git://|https://|' -e 's/\.git$//')
      echo "source=$source" >> $GITHUB_OUTPUT

      echo "revision=$REVISION" >> $GITHUB_OUTPUT
```

stefanprodan reviewed (approved):
LGTM

Awesome job thanks @matheuscscp
